A40-csds-support.md

```diff
@@ -0,0 +1,328 @@
+proposal. But the service itself has the potential to serve client status on xDS
+clients. However, during development, we found several constraints about the
+existing service protocol, hence we merged several updates (see above) to
+improve the CSDS service to meet our standard. The updates are made to xDS v3
+and xDS v4 (alpha) only, since xDS v2 is in deprecated state. This doc proposes
+to not support xDS v2 for CSDS.
+
+#### Detail: Expose xDS Config As Is
+
+To date, gRPC supports the majority of xDS config but not all. Alternatively, we
+could only dump the config that gRPC understands. However, doing so will create
+a behavior difference between the gRPC and the control plane. It might be
+confusing to users that some fields are missing. Also, during the configuration
+dump, Envoy also dumps its entire configuration even if it doesn’t understand
+some of the configuration fields (think of fields for Envoy extensions). So, for
+correctness and simplicity, we should **cache and expose the xDS config the way
+gRPC receives them**.
+
+#### Detail: Cache Lifecycle
+
+gRPC doesn't use xDS messages directly, but interprets them into language-native
+class/struct. The lifecycle of the cached xDS messages should be identical to
+the interpreted structure in each stack, so there is no memory management logic
+change needed.
+
+#### Detail: Generated Node Information
+
+gRPC loads Node information from a [bootstrap
+file](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md#xdsclient-and-bootstrap-file),
+which includes the identification that will be used in the control plane. gRPC's
+xDS client also generates `user_agent_name`, `user_agent_version`, and
+`client_features` before sending the `Node` information to the control plane.
+This information can be very helpful to locate the culprit release and
+debugging. This doc recommends the implementation to provide the
+`ClientStatusResponse.ClientConfig.node` field with the `Node` information that
+the control plane will receive. Here is an example of a filled `Node`:
+
+```json
+{
+  "node": {
+    "id": "c591240a-b3b6-4761-aada-3fa43ec6a852",
+    "cluster": "cluster",
+    "metadata": {
+      "...": "..."
+    },
+    "locality": {
+      "zone": "zone"
+    },
+    "userAgentName": "gRPC Java",
+    "userAgentVersion": "1.35.1-SNAPSHOT",
+    "clientFeatures": ["envoy.lb.does_not_support_overprovisioning"]
+  }
+}
+```
+
+#### Detail: Node Matching
+
+The node matching mechanism is not designed for gRPC’s use case, but for
+querying the control plane. Technically, gRPC application only serves as one xDS
+node, so there is no need to differentiate with xDS client's status to reply.
+However, gRPC should reject the request with Node Matcher that failed to match
+with the Node information of the gRPC application. This feature has lower
+priority. To preserve forward compatibility, this doc proposes to throw an
```

Comment:
s/throw/return/

(This is not an exception, it's a status code.)

Reply:
Done.

(In Python, it is an exception)